source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,264 @@
+// key/priority. Also, this should be part of some utility library.
+// https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html
+bool GzipFilter::isAcceptEncodingAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* accept_encoding = headers.AcceptEncoding();
+
+  if (accept_encoding) {
+    bool is_wildcard = false; // true if found and not followed by `q=0`.
+
+    for (const auto token : StringUtil::splitToken(headers.AcceptEncoding()->value().c_str(), ",",
+                                                   false /* keep_empty */)) {
+      const auto value =
+          StringUtil::trim(StringUtil::cropRight(token, ";", false /* trim_whitespace */));
+      const auto q_value =
+          StringUtil::trim(StringUtil::cropLeft(token, ";", false /* trim_whitespace */));
+
+      // If value is the gzip coding, check the qvalue and return.
+      if (value == Http::Headers::get().AcceptEncodingValues.Gzip) {
+        return (q_value != ZeroQvalueString);
+      }
+
+      // If value is the identity coding, just check the qvalue and return.
+      if (value == Http::Headers::get().AcceptEncodingValues.Identity) {
+        return (q_value != ZeroQvalueString);
+      }
+
+      if (value == Http::Headers::get().AcceptEncodingValues.Wildcard) {
+        is_wildcard = (q_value != ZeroQvalueString);
+      }
+    }
+
+    // if neither identity nor gzip codings are present, we return the wildcard.
+    return is_wildcard;
+  }
+
+  return true;
+}
+
+bool GzipFilter::isContentTypeAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* content_type = headers.ContentType();
+  if (content_type && !config_->contentTypeValues().empty()) {
+    std::string value{
+        StringUtil::trim(StringUtil::cropRight(content_type->value().c_str(), ";", false))};
+    return config_->contentTypeValues().find(value) != config_->contentTypeValues().end();
+  }
+  return true;
+}
+
+bool GzipFilter::isEtagAllowed(HeaderMap& headers) const {
+  return !(config_->disableOnEtagHeader() && headers.Etag());
+}
+
+bool GzipFilter::isMinimumContentLength(HeaderMap& headers) const {
+  const Http::HeaderEntry* content_length = headers.ContentLength();
+  if (content_length) {
+    uint64_t length;
+    return StringUtil::atoul(content_length->value().c_str(), length) &&
+           length >= config_->minimumLength();
+  }
+
+  const Http::HeaderEntry* transfer_encoding = headers.TransferEncoding();
+  return (transfer_encoding &&
+          StringUtil::findToken(transfer_encoding->value().c_str(), ",",
+                                Http::Headers::get().TransferEncodingValues.Chunked.c_str()));
+}
+
+bool GzipFilter::isTransferEncodingAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* transfer_encoding = headers.TransferEncoding();
+  if (transfer_encoding) {
+    for (auto header_value :
+         StringUtil::splitToken(transfer_encoding->value().c_str(), ",", true)) {
```

Comment:
could you add HeaderMap::string_view()  -- it  stores the length in .string_length_ but here we have to rediscover the length with a strlen in the string_view ctor.  You could add this as a TODO now and cover it in another pass as I don't want you make this PR bigger.  There are probably other places where this would be useful as well in the code.

Reply:
SGTM 👍